topdown/graphql.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2022 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+	gqltop "github.com/vektah/gqlparser/v2"
+	gqlast "github.com/vektah/gqlparser/v2/ast"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+	gqlparser "github.com/vektah/gqlparser/v2/parser"
+	gqlvalidator "github.com/vektah/gqlparser/v2/validator"
+)
+
+func errGraphQLError(builtinName, msg string) error {
+	return fmt.Errorf("%s error: %s", builtinName, msg)
+}
+
+// Generate a useful error message from a GraphQL error list.
+// Modeled after the approaches used for similar error list types
+// in docker/containerd.
+func errGraphQLErrorList(builtinName string, errs gqlerror.List) error {
+	switch len(errs) {
+	case 0:
+		return fmt.Errorf("graphql internal error:%s: Error list invoked with zero arguments", builtinName)
+	case 1:
+		return errs[0].Unwrap()
+	default:
+		msg := "errors:\n"
+		for _, err := range errs {
+			msg += builtinName + " error:" + err.Error() + "\n"
+		}
+		return fmt.Errorf(msg)
+	}
+}
+
+// This is a "show only the first error" approach, but under normal
+// conditions, ast.Error structs should only have one location for each
+// error.
+// Note: This means a caller with unusual syntax errors may have to
+// go through the parse/debug cycle multiple times in the worst case to
+// find all of their issues.
+// TODO: Extend this to handle error lists correctly.
+func errGraphQLASTError(builtinName string, e gqlerror.Error) error {
+	return errGraphQLError(builtinName, fmt.Sprintf("%s line:%d,col:%d", e.Message, e.Locations[0].Line, e.Locations[0].Column))
+}
+
+func builtinGraphQLParse(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Get the raw strings from each operand so that we can
+	// feed them to the GraphQL parser functions.
+	raw_query, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+	raw_schema, err := builtins.StringOperand(operands[1].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Generate ASTs/errors for the GraphQL schema and query.
+	schema, errS := gqltop.LoadSchema(&gqlast.Source{Input: string(raw_schema)})
+	if errS != nil {
+		return errS.Unwrap()
```

Comment:
It is indeed a wrappered error type! The underlying [`gqlparser` library's `Error` type](https://github.com/vektah/gqlparser/blob/d3d9eb00157553f2203376295a8686862c41614a/gqlerror/error.go#L13) is a struct that wrappers over the underlying `error`:

```go
type Error struct {
	err        error                  `json:"-"`
	Message    string                 `json:"message"`
	Path       ast.Path               `json:"path,omitempty"`
	Locations  []Location             `json:"locations,omitempty"`
	Extensions map[string]interface{} `json:"extensions,omitempty"`
	Rule       string                 `json:"-"`
}
```
The underlying `error` can then be extracted by calling `Unwrap()`, which just returns the `err` field directly.

The best approach I could see at the time was to deal with these `gqlerror.Error` structs specially, and if non-nil, `Unwrap` them for returning to the caller.

Do you think there's a cleaner way of accomplishing the same thing? (This may tie into the "structured error messages" comment you made earlier. :slightly_smiling_face: )

Reply:
Implemented the approach you suggested in an [earlier comment](https://github.com/open-policy-agent/opa/pull/4684#discussion_r874710677). Thanks for the suggestion! :smiley: